components/ai_chat/core/ai_chat_helper.cc

```diff
@@ -0,0 +1,586 @@
+    VLOG(1) << __func__ << ": No data";
+    return;
+  }
+
+  // tokens + max_new_tokens must be <= 4096 (llama2)
+  // 8092 chars, ~3,098 tokens (reserved for article)
+  // 1k chars, ~380 tokens (reserved for prompt)
+  contents_text = contents_text.substr(0, 8092);
+
+  is_video_ = is_video;
+  article_text_ = contents_text;
+  engine_->SanitizeInput(article_text_);
+
+  OnPageHasContentChanged();
+
+  // Now that we have article text, we can suggest to summarize it
+  DCHECK(suggested_questions_.empty())
+      << "Expected suggested questions to be clear when there has been no"
+      << " previous text content but there were " << suggested_questions_.size()
+      << " suggested questions: "
+      << base::JoinString(suggested_questions_, ", ");
+
+  // Now that we have content, we can provide a summary on-demand. Add that to
+  // suggested questions.
+  // TODO(petemill): translation for this question
+  suggested_questions_.emplace_back(is_video_ ? "Summarize this video"
+                                              : "Summarize this page");
+  OnSuggestedQuestionsChanged();
+  MaybeGenerateQuestions();
+}
+
+void AIChatHelper::CleanUp() {
+  chat_history_.clear();
+  article_text_.clear();
+  suggested_questions_.clear();
+  pending_request_.reset();
+  is_page_text_fetch_in_progress_ = false;
+  is_request_in_progress_ = false;
+  has_generated_questions_ = false;
+  should_page_content_be_disconnected_ = false;
+  OnSuggestedQuestionsChanged();
+  SetAPIError(mojom::APIError::None);
+  engine_->ClearAllQueries();
+
+  // Trigger an observer update to refresh the UI.
+  for (auto& obs : observers_) {
+    obs.OnHistoryUpdate();
+    obs.OnPageHasContent();
+  }
+}
+
+int64_t AIChatHelper::GetNavigationId() const {
+  return current_navigation_id_;
+}
+
+void AIChatHelper::SetNavigationId(int64_t navigation_id) {
+  current_navigation_id_ = navigation_id;
+}
+
+bool AIChatHelper::IsSameDocumentNavigation() const {
+  return is_same_document_navigation_;
+}
+
+void AIChatHelper::SetSameDocumentNavigation(bool same_document_navigation) {
+  is_same_document_navigation_ = same_document_navigation;
+}
+
+std::vector<std::string> AIChatHelper::GetSuggestedQuestions(
+    bool& can_generate,
+    mojom::AutoGenerateQuestionsPref& auto_generate) {
+  // Can we get suggested questions
+  can_generate = !has_generated_questions_ && !article_text_.empty();
+  // Are we allowed to auto-generate
+  auto_generate = GetAutoGeneratePref();
+  return suggested_questions_;
+}
+
+bool AIChatHelper::HasPageContent() {
+  return !article_text_.empty();
+}
+
+void AIChatHelper::DisconnectPageContents() {
+  CleanUp();
+
+  should_page_content_be_disconnected_ = true;
+}
+
+void AIChatHelper::ClearConversationHistory() {
+  chat_history_.clear();
+  engine_->ClearAllQueries();
+
+  // Trigger an observer update to refresh the UI.
+  for (auto& obs : observers_) {
+    obs.OnHistoryUpdate();
+  }
+}
+
+mojom::APIError AIChatHelper::GetCurrentAPIError() {
```

Comment:
This is not a great pattern and should be addressed in a follow-up. API methods should have callbacks to handle errors. Checking this is very unreliable, especially if you have more than one api call happening at the same time.

Reply:
commented on the original PR https://github.com/brave/brave-core/pull/19804/files#r1349226241

Reply:
I responded over there, I think it's fine if not perfect.